text/0038-output-types-for-tokenization-and-sc/0038-output-types-for-tokenization-and-sc.md

```diff
@@ -0,0 +1,2994 @@
+corresponding to that index. Whenever a new milestone appears, nodes perform the white-flag ordering and transaction
+validation on its past cone. The timestamp and milestone index of the confirming milestone provide the time as an input
+parameter to transaction validation.
+
+An output that contains a <i>Timelock Block</i> can not be unlocked before the specified timelock has expired. The
+timelock is expired when the timestamp and/or milestone index of the confirming milestone is equal or past the
+timestamp and/or milestone defined in the <i>Timelock Block</i>.
+
+The timelock can be specified as a unix timestamp or as a milestone index. When specified in both ways, both conditions
+have to pass in order for the unlock to be valid.
+
+The two time representations help to protect against the possible downtime of the Coordinator. If the Coordinator is
+down, "milestone index clock" essentially stops advancing, while "real time clock" does not. An output that specifies
+time in both clocks must satisfy both conditions (AND relation).
+
+##### Additional semantic transaction validation rules:
+- An output that has <i>Timelock Milestone Index Block</i> specified must only be consumed and unlocked in a
+  transaction, if the confirming milestone index is ≥ than the `Milestone Index` specified in the block.
+- An output that has <i>Timelock Unix Block</i> specified must only be consumed and unlocked in a transaction, if the
+  timestamp of the confirming milestone is equal or past the `Unix Time` specified in the block.
+- An output that has both <i>Timelock Milestone Index Block</i> and  <i>Timelock Unix Block</i> specified must only be
+  consumed and unlocked in a transaction if both blocks validate.
+
+<details>
+<summary>Timelock Milestone Index Block</summary>
+<blockquote>
+    Defines a milestone index until which the output can not be unlocked.
+</blockquote>
+</details>
+
+<table>
+    <tr>
+        <td><b>Name</b></td>
+        <td><b>Type</b></td>
+        <td><b>Description</b></td>
+    </tr>
+    <tr>
+        <td>Block Type</td>
+        <td>uint8</td>
+        <td>
+            Set to <strong>value 3</strong> to denote a <i>Timelock Milestone Index Block</i>.
+        </td>
+    </tr>
+    <tr>
+        <td>Milestone Index</td>
+        <td>uint32</td>
+        <td>
+            The milestone index starting from which the output can be consumed.
+        </td>
+    </tr>
+</table>
+
+<details>
+<summary>Timelock Unix Block</summary>
+<blockquote>
+    Defines a unix time until which the output can not be unlocked.
+</blockquote>
+</details>
+
+<table>
+    <tr>
+        <td><b>Name</b></td>
+        <td><b>Type</b></td>
+        <td><b>Description</b></td>
+    </tr>
+    <tr>
+        <td>Block Type</td>
+        <td>uint8</td>
+        <td>
+            Set to <strong>value 4</strong> to denote a <i>Timelock Unix Block</i>.
+        </td>
+    </tr>
+    <tr>
+        <td>Unix Time</td>
+        <td>uint32</td>
```

Comment:
```suggestion
        <td>uint64</td>
```

Reply:
It has been considered already and the decision was to go with uint32.
With that, we are fine until February 7, 2106, while sparing 4 additional bytes per time value in outputs that would be zero for the next 80+ years anyway. Let's set a reminder for 2100 not to forget to update then :smile_cat:

ps: Bitcoin also uses [uint32 block timestamps](https://en.bitcoin.it/wiki/Protocol_documentation).

Reply:
Fair enough! I stumbled upon this and thought I would point it out. I figured it would make sense to be consistent with the [milestone payload](https://github.com/iotaledger/protocol-rfcs/blob/master/text/0019-milestone-payload/0019-milestone-payload.md#structure) which uses `uint64` for the timestamp (at least in the RFC). But I guess 2106 seems fine :smile:.